src/consensus/block_producer.rs

```diff
@@ -372,22 +370,24 @@ where
                     .await;
             }
         }
-        if is_last {
-            Ok(Some(maybe_block_hash.unwrap()))
-        } else {
-            assert!(maybe_block_hash.is_none());
-            Ok(None)
-        }
+        assert_eq!(maybe_block_hash.is_some(), is_last);
+        Ok(maybe_block_hash)
     }
 }
 
-// TODO: extend docstring
-/// Returns
+/// Produces a slice (payload only).
+///
+/// Packs transactions received from the network into the slice until either:
+/// - The slice is full (accounting for encoding of parent), OR
+/// - there is an incoming message on the `preempt_receiver`.
+///
+/// Returns the slice payload, amount of time elapsed, and number of transactions packed.
 async fn produce_slice_payload<T>(
```

Comment:
I am wondering if having two versions of this function would make them simpler albeit with some code duplication.

Parent ready version:
`produce_slice_payload_parent_ready(txs_receiver, parent:Option<BlockId>, duration) -> (SlicePayload, Duration, usize)`: it doesn't need a preemption channel or a minimum duration.  It needs a maximum duration.  

Parent not ready version: would probably not change very much but it we will have a simpler to understand more straightforward one above.